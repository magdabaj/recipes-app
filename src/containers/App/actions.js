import {LOAD_RECIPES_ERROR, LOAD_TAGS, LOAD_TAGS_ERROR, LOAD_TAGS_SUCCESS} from "./constants";

// export const loadRecipes = page => ({
//     type: LOAD_RECIPES,
//     page,
// });
//
// export const loadRecipesSuccess = response => ({
//     type: LOAD_RECIPES_SUCCESS,
//     response,
// });

export const loadRecipesError = error => ({
    type: LOAD_RECIPES_ERROR,
    error,
});

export const loadTags = () => ({
    type: LOAD_TAGS,
});

export const loadTagsSuccess = tags => ({
    type: LOAD_TAGS_SUCCESS,
    tags,
});

export const loadTagsError = error => ({
    type: LOAD_TAGS_ERROR,
    error,
});