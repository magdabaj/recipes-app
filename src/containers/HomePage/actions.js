/*
 *
 * RecipesHomePage actions
 *
 */

import {
    CHANGE_PAGE,
    CHANGE_PAGE_ERROR,
    CHANGE_PAGE_SUCCESS,
    LOAD_RECIPES,
    LOAD_RECIPES_BY_TAG,
    LOAD_RECIPES_BY_TAG_ERROR,
    LOAD_RECIPES_BY_TAG_SUCCESS,
    LOAD_RECIPES_ERROR,
    LOAD_RECIPES_SUCCESS,
    LOAD_TAGS,
    LOAD_TAGS_ERROR,
    LOAD_TAGS_SUCCESS,
} from './constants';

export const loadRecipes = page => ({
    type: LOAD_RECIPES,
    page,
});

export const loadRecipesSuccess = response => ({
    type: LOAD_RECIPES_SUCCESS,
    response,
});

export const loadRecipesError = error => ({
    type: LOAD_RECIPES_ERROR,
    error,
});

export const loadRecipesByTag = (tagId, page) => ({
    type: LOAD_RECIPES_BY_TAG,
    tagId,
    page,
});

export const loadRecipesByTagSuccess = response => ({
    type: LOAD_RECIPES_BY_TAG_SUCCESS,
    response,
});

export const loadRecipesByTagError = error => ({
    type: LOAD_RECIPES_BY_TAG_ERROR,
    error,
});

// export const changePage = page => ({
//   type: CHANGE_PAGE,
//   page,
// });
//
// export const changePageSuccess = response => ({
//   type: CHANGE_PAGE_SUCCESS,
//   response,
// });
//
// export const changePageError = error => ({
//   type: CHANGE_PAGE_ERROR,
//   error,
// })
//
