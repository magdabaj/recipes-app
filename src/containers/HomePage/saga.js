import { take, call, put, select, takeLatest, all } from 'redux-saga/effects';
import {
    loadRecipesByTagError,
    loadRecipesByTagSuccess,
    loadRecipesError,
    loadRecipesSuccess,
} from './actions';
import { LOAD_RECIPES, LOAD_RECIPES_BY_TAG } from './constants';
import { fetchAllRecipes, fetchRecipesByTag } from '../../utils/api/recipe';

export function* loadRecipesSaga(action) {
    const { page } = action;
    try {
        const response = yield call(fetchAllRecipes, page);
        yield put(loadRecipesSuccess(response));
    } catch (error) {
        yield put(loadRecipesError(error.message));
    }
}

export function* watchRecipesLoad() {
    yield takeLatest(LOAD_RECIPES, loadRecipesSaga);
}

export function* loadRecipesByTagSaga(action) {
    const { tagId, page } = action;

    try {
        const response = yield call(fetchRecipesByTag, tagId, page);
        yield put(loadRecipesByTagSuccess(response));
    } catch (e) {
        yield put(loadRecipesByTagError(e.message));
    }
}

export function* watchRecipesByTagLoad() {
    yield takeLatest(LOAD_RECIPES_BY_TAG, loadRecipesByTagSaga);
}