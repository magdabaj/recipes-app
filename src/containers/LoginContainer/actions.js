/*
 *
 * LoginFormContainer actions
 *
 */

import {
    AUTHENTICATE,
    AUTHENTICATE_ERROR,
    AUTHENTICATE_SUCCESS, SET_ERROR,
    SIGN_IN,
    SIGN_IN_ERROR,
    SIGN_IN_SUCCESS,
    SIGN_OUT,
    SIGN_OUT_ERROR,
    SIGN_OUT_SUCCESS,
    SIGN_UP_USER,
    SIGN_UP_USER_ERROR,
    SIGN_UP_USER_SUCCESS
} from './constants'

export const setError = () => ({
    type: SET_ERROR
})

export const signUp = ({ email, password }) => ({
    type: SIGN_UP_USER,
    email,
    password,
});

export const signUpSuccess = (json, email) => ({
    type: SIGN_UP_USER_SUCCESS,
    ...json,
    email,
});

export const signUpError = error => ({
    type: SIGN_UP_USER_ERROR,
    error,
});

export const signOut = () => ({
    type: SIGN_OUT,
});

export const signOutError = error => ({
    type: SIGN_OUT_ERROR,
    error,
});

export const signOutSuccess = json => ({
    type: SIGN_OUT_SUCCESS,
    ...json,
});

export const signIn = ({ email, password }) => ({
    type: SIGN_IN,
    email,
    password,
});

export const signInSuccess = (json, email) => ({
    type: SIGN_IN_SUCCESS,
    ...json,
    email,
});

export const signInError = error => ({
    type: SIGN_IN_ERROR,
    error,
});

export const authenticate = () => ({
    type: AUTHENTICATE,
});

export const authenticateSuccess = json => ({
    type: AUTHENTICATE_SUCCESS,
    ...json,
});

export const authenticateError = error => ({
    type: AUTHENTICATE_ERROR,
    error,
});
