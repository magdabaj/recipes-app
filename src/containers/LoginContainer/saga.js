import { call, put, takeLatest } from 'redux-saga/effects';
import { AUTHENTICATE, SIGN_IN, SIGN_OUT, SIGN_UP_USER } from './constants';
import { signInUser, signUpUser, signOutUser, authenticateUser } from '../../utils/api/user';
import {
    authenticateError,
    authenticateSuccess, clearLogoutStatus, clearStatus,
    signInError,
    signInSuccess,
    signOutError,
    signOutSuccess,
    signUpError,
    signUpSuccess,
} from './actions';

export function* handleSignUp(action) {
    const { email, password } = action;
    try {
        const json = yield call(signUpUser, email, password);
        if (json.type === 'error') {
            yield put(signUpError(json.message));
            yield put(clearStatus());
        } else {
            yield put(signUpSuccess(json, email));
            yield put(clearStatus());
        }
    } catch (e) {
        yield put(signUpError(e.message));
        yield put(clearStatus());
    }
}

export function* watchSignUp() {
    yield takeLatest(SIGN_UP_USER, handleSignUp);
}

export function* handleSignIn(action) {
    const { email, password } = action;
    try {
        const json = yield call(signInUser, email, password);
        if (json.type === 'error') {
            yield put(signInError(json.message));
            yield put(clearStatus());
        } else {
            yield put(signInSuccess(json, email));
            yield put(clearStatus())
        }
    } catch (e) {
        yield put(signInError(e.message));
    }
}

export function* watchSignIn() {
    yield takeLatest(SIGN_IN, handleSignIn);
}

export function* handleSignOut() {
    try {
        const json = yield call(signOutUser);
        if (json.type === 'error') {
            yield put(signOutError(json.message));
            yield put(clearLogoutStatus());
        } else {
            yield put(signOutSuccess(json));
            yield put(clearLogoutStatus());
        }
    } catch (e) {
        yield put(signOutSuccess(e.message));
        yield put(clearLogoutStatus());
    }
}

export function* watchSignOut() {
    yield takeLatest(SIGN_OUT, handleSignOut);
}

export function* handleAuthenticate() {
    try {
        const json = yield call(authenticateUser);
        if (json.type === 'error') {
            yield put(authenticateError(json.message));
        } else {
            yield put(authenticateSuccess(json));
            yield put(clearStatus());
        }
    } catch (e) {
        yield put(authenticateError(e.message));
        yield put(clearStatus());
    }
}

export function* watchAuthenticate() {
    yield takeLatest(AUTHENTICATE, handleAuthenticate);
}
