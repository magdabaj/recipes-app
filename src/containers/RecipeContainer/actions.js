/*
 *
 * RecipeContainer actions
 *
 */

import {
    GET_RATINGS_ERROR,
    GET_RATINGS,
    GET_RATINGS_SUCCESS,
    GET_RECIPE,
    GET_RECIPE_ERROR,
    LOAD_TAGS,
    LOAD_TAGS_SUCCESS,
    LOAD_TAGS_ERROR,
    GET_RECIPE_SUCCESS,
    SEND_RATING,
    SEND_RATING_ERROR,
    SEND_RATING_SUCCESS,
    GET_COMMENTS,
    GET_COMMENTS_SUCCESS,
    GET_COMMENTS_ERROR,
    ADD_COMMENT,
    ADD_COMMENT_SUCCESS,
    ADD_COMMENT_ERROR,
    EDIT_COMMENT,
    EDIT_COMMENT_SUCCESS,
    EDIT_COMMENT_ERROR,
    REMOVE_COMMENT,
    REMOVE_COMMENT_SUCCESS,
    REMOVE_COMMENT_ERROR, CLEAR_STATUS, CLEAR_DELETE_COMMENTS_STATUS,
} from './constants';

export const getRecipeRatings = recipeId => ({
    type: GET_RATINGS,
    recipeId,
});

export const getRecipeRatingsSuccess = (message, recipeId, ratingsMean) => ({
    type: GET_RATINGS_SUCCESS,
    message,
    recipeId,
    ratingsMean,
});

export const getRecipeRatingsError = error => ({
    type: GET_RATINGS_ERROR,
    error,
});

export const getRecipe = recipeId => ({
    type: GET_RECIPE,
    recipeId,
});

export const getRecipeSuccess = recipe => ({
    type: GET_RECIPE_SUCCESS,
    recipe,
});

export const getRecipeError = error => ({
    type: GET_RECIPE_ERROR,
    error,
});


export const sendRating = (rate, userId, recipeId) => ({
    type: SEND_RATING,
    rate,
    userId,
    recipeId,
});

export const sendRatingSuccess = (message, recipeId, ratingsMean) => ({
    type: SEND_RATING_SUCCESS,
    message,
    recipeId,
    ratingsMean,
});

export const sendRatingError = error => ({
    type: SEND_RATING_ERROR,
    error,
});

export const getComments = recipeId => ({
    type: GET_COMMENTS,
    recipeId,
});

export const getCommentsSuccess = response => ({
    type: GET_COMMENTS_SUCCESS,
    response,
});

export const getCommentsError = error => ({
    type: GET_COMMENTS_ERROR,
    error,
});

export const addComment = (recipeId, content, userId) => ({
    type: ADD_COMMENT,
    recipeId,
    userId,
    content,
});

export const addCommentSuccess = message => ({
    type: ADD_COMMENT_SUCCESS,
    message,
});

export const addCommentError = error => ({
    type: ADD_COMMENT_ERROR,
    error,
});

export const editComment = (commentId, content, userId) => ({
    type: EDIT_COMMENT,
    commentId,
    userId,
    content,
});

export const editCommentSuccess = message => ({
    type: EDIT_COMMENT_SUCCESS,
    message,
});

export const editCommentError = error => ({
    type: EDIT_COMMENT_ERROR,
    error,
});

export const removeComment = commentId => ({
    type: REMOVE_COMMENT,
    commentId,
});

export const removeCommentSuccess = message => ({
    type: REMOVE_COMMENT_SUCCESS,
    message,
});

export const removeCommentError = error => ({
    type: REMOVE_COMMENT_ERROR,
    error,
});

export const clearStatus = () => ({
    type: CLEAR_STATUS
})

export const clearDeleteCommentsStatus = () => ({
    type: CLEAR_DELETE_COMMENTS_STATUS
})
