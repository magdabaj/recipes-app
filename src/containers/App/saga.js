import {call, put, takeLatest} from "redux-saga/effects";
import {fetchAllTags} from "../../utils/api/tags";
import {loadTagsError, loadTagsSuccess} from "./actions";
import {LOAD_TAGS} from "./constants";

export function* loadTagsSaga() {
    try {
        const tags = yield call(fetchAllTags);
        yield put(loadTagsSuccess(tags));
    } catch (error) {
        yield put(loadTagsError(error.message));
    }
}

export function* watchTagsLoad() {
    yield takeLatest(LOAD_TAGS, loadTagsSaga);
}