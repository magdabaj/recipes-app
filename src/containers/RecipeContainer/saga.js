import { all, call, put, select, takeLatest } from 'redux-saga/effects';

import { createRating, fetchRecipeRatings } from '../../utils/api/rating';
import {
    getRecipeError,
    getRecipeRatingsError,
    getRecipeRatingsSuccess,
    getRecipeSuccess,
    sendRatingSuccess,
    sendRatingError,
    getCommentsError,
    getCommentsSuccess,
    addCommentError,
    addCommentSuccess,
    getComments,
    removeCommentSuccess,
    removeCommentError, clearStatus,
} from './actions';
import {
    ADD_COMMENT,
    GET_COMMENTS,
    GET_RATINGS,
    GET_RECIPE,
    REMOVE_COMMENT,
    SEND_RATING,
} from './constants';
import { fetchRecipe } from '../../utils/api/recipe';
import { deleteComment, fetchComments, storeComment } from '../../utils/api/comment';
import { makeSelectUser } from '../LoginContainer/selectors';

export function* handleRecipeRatingsLoad(action) {
    const { recipeId } = action;

    try {
        const { message, ratingsMean } = yield call(fetchRecipeRatings, recipeId);
        yield put(getRecipeRatingsSuccess(message, recipeId, ratingsMean));
    } catch (e) {
        yield put(getRecipeRatingsError(e.message));
    }
}

export function* watchRecipeRatingsLoad() {
    yield takeLatest(GET_RATINGS, handleRecipeRatingsLoad);
}

export function* handleRatingSend(action) {
    const { rate, userId, recipeId } = action;
    try {
        const json = yield call(createRating, rate, userId, recipeId);
        if (json.type === 'error') {
            yield put(sendRatingError(json.message));
        } else {
            yield put(sendRatingSuccess(json, recipeId, json.ratingsMean));
        }
    } catch (e) {
        yield put(sendRatingError(e.message));
    }
}

export function* watchRatingSend() {
    yield takeLatest(SEND_RATING, handleRatingSend);
}

function* handleRecipeLoad(action) {
    const { recipeId } = action;
    try {
        const recipe = yield call(fetchRecipe, recipeId);
        yield put(getRecipeSuccess(recipe));
    } catch (e) {
        yield put(getRecipeError(e.message));
    }
}

export function* watchRecipeLoad() {
    yield takeLatest(GET_RECIPE, handleRecipeLoad);
}

export function* loadCommentsSaga({ recipeId }) {
    try {
        const comments = yield call(fetchComments, recipeId);
        yield put(getCommentsSuccess(comments));
    } catch (e) {
        yield put(getCommentsError(e.message));
    }
}

export function* watchCommentsLoad() {
    yield takeLatest(GET_COMMENTS, loadCommentsSaga);
}

export function* addCommentSaga({ recipeId }) {
    const { userId } = yield select(makeSelectUser());
    try {
        const message = yield call(storeComment, recipeId.recipeId, recipeId.content, userId);
        yield put(addCommentSuccess(message));
        yield put(clearStatus());
    } catch (e) {
        yield put(addCommentError(e.message));
        yield put(clearStatus());
    }
}

export function* watchCommentAdd() {
    yield takeLatest(ADD_COMMENT, addCommentSaga);
}

export function* deleteCommentSaga(action) {
    const { commentId } = action;
    const { userId } = yield select(makeSelectUser());

    try {
        const message = yield call(deleteComment, commentId, userId);
        yield put(removeCommentSuccess(message));
        yield put(clearStatus());
    } catch (e) {
        yield put(removeCommentError(e.message));
        yield put(clearStatus());
    }
}

export function* watchCommentDelete() {
    yield takeLatest(REMOVE_COMMENT, deleteCommentSaga);
}

