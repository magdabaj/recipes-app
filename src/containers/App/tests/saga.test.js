import {put} from "redux-saga/effects";
import {loadTagsError, loadTagsSuccess} from "../actions";
import {loadTagsSaga} from '../saga'
import {beforeEach, describe, expect} from "@jest/globals";

describe('loadTags saga', () => {
    let loadTagsGenerator;

    beforeEach(() => {
        loadTagsGenerator = loadTagsSaga();

        const selectDescriptor = loadTagsGenerator.next().value;
        expect(selectDescriptor).toMatchSnapshot();
    })

    it('should dispatch the loadTagsSuccess action if it requests the data successfully', function () {
        const response = [
            {
                id: 1,
            },
            {
                id: 2,
            }
        ]
        const putDescriptor = loadTagsGenerator.next(response).value;
        expect(putDescriptor).toEqual(put(loadTagsSuccess(response)))
    });

    it('should dispatch the loadTagsError action if the request fails', function () {
        const response = new Error('some error');
        const putDescriptor = loadTagsGenerator.throw(response).value;
        expect(putDescriptor).toEqual(put(loadTagsError(response.message)))
    });
})