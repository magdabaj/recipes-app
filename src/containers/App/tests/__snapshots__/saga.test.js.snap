// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadTags saga should dispatch the loadTagsError action if the request fails 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`loadTags saga should dispatch the loadTagsSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
